### ISSUE TITLE
NER Model retrain

### ISSUE THREAD

**Issue by AIRLegend** [explosion/spaCy#3059]
Hi. Sorry for the inconvenience. I already posted my problem on StackOverflow and I had no awnser.
https://stackoverflow.com/questions/53806376/spacyv2-ner-model-retrain

## Problem:

I'm trying to retrain the Spanish spacy model to recognize a new entity type (PAIN). I'm training the model with a handmade corpus of 70 phrases (I know that is not enough but my purpose is only didactic).

The problem is that after training it for more than 2 epochs, it doesn't even recognize the default entities in another phrase let alone my custom entity in the training set, so I don't think its a *forgetting problem*.

```
nlp = spacy.load('es_core_news_sm')
if 'ner' not in nlp.pipe_names:
    ner = nlp.create_pipe('ner')
    nlp.add_pipe(ner)
else:
    ner = nlp.get_pipe('ner')

label = 'PAIN'
matches = ['doler', 'duelen', 'dolor', 'duele', 'dolores', 'dolia', 'dolian']
matcher = PhraseMatcher(nlp.vocab)

for i in matches:
    # adding the label
    matcher.add(label, None, nlp(i))

res = []
to_train_ents = []
with open('../data/train_ent_pain.txt') as dt: #contains de training phrases
    line = True
    while line:
        line = dt.readline()
        mnlp_line = nlp(line.rstrip())
        matches = matcher(mnlp_line)
        res = [offseter(label, mnlp_line, x) for x in matches] #Convert the offsets with a custom function
        to_train_ents.append((line.rstrip(), dict(entities=res)))

optimizer = nlp.begin_training()
other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']

ner.add_label(label)

with nlp.disable_pipes(*other_pipes): #only train NER
    for itn in range(15):

        random.shuffle(to_train_ents)
        losses = {}
        # batch up the examples using spaCy's minibatch
        batches = minibatch(to_train_ents, size=compounding(4., 32., 1.001))

        for batch in batches:
            texts, annotations = zip(*batch)
            nlp.update(texts, annotations, sgd=optimizer, drop=0.35, losses=losses)
        print('{} of {} Losses: {}'.format(itn, 15, losses))
```

When I test the model with the same train data it recognizes nothing (prints empty arrays) and even with phrases which work before the training.

```
mnlp_line = nlp('Apple es una compañía que opera en la bolsa de Estados Unidos.')
ents = [(x.text, x.label_) for x in mnlp_line.ents]
print(ents)
```

## Your Environment

* Operating System: MacOS Mojave
* Python Version Used: 3.6
* spaCy Version Used: v2

Thank you in advance and sorry again for posting here this.

**Comment by ines** [comment]
Try running your code without `optimizer = nlp.begin_training()` and without passing the optimizer into `nlp.update`!

`nlp.begin_training` will reset the weights and initialize them randomyly, which is usually what you want if you're training a new model from scratch. But if you're updating an existing model, this can lead to unintended side-effects like the one you're experiencing. I've also just updated the NER training example with a comment that reflects this!

**Comment by AIRLegend** [comment]
Just tried that with the same result  😕. Also tried with the other Spanish model provided.

**Comment by honnibal** [comment]
Hmm.

You end up with two NER models in the pipeline there, which I think might be your problem. After the first model, your entity words are already marked as not entities, so your second model doesn't have a chance to make a prediction. I'm not 100% sure this is what's going on, but it seems likely.

We don't really have a super convenient solution for this at the moment, because the `token.ent_iob` property is not exposed as writeable (as setting individual values can cause inconsistent data, which can cause segfaults that are hard to debug). The best solution would be to use `doc.to_array([ENT_IOB])` (importing `ENT_IOB` from `spacy.attrs`. The numeric codes for the ent IOB values go:

* 0: Unset
* 1: In (i.e. within an entity)
* 2: Out: (i.e. not an entity word)
* 3: Begin: (i.e. the beginning of an entity)

You want to take all the 2 values (the O's) and set them to 0. This will preserve the entities preset by the previous model, while allowing your second model to predict over the words that the first model marked as O.

**Comment by AIRLegend** [comment]
If I understood correctly, when predicting:
```
mnlp_line = nlp('Apple es una compañía que opera en la bolsa de Estados Unidos.')
ar = mnlp_line.to_array([ENT_IOB])
ar[ar == 2] = 0
mnlp_line = mnlp_line.from_array([ENT_IOB], ar.reshape((ar.shape[0],1)))
ents = [(x.text, x.label_) for x in mnlp_line.ents]
print(ents)
``` 
( I've made that also with my training phrases).
Still gives me no entities

**Comment by AIRLegend** [comment]
Just tested the example provided but loading my corpus in `TRAIN_DATA`. 
It regoznices no entities when trying to add new categories to the model, but it works when training new entities on the existing labels (as in the example).

**Comment by honnibal** [comment]
Your code for the to/from array looks good, but you need to do that after the predictions of the first model, and before your second model runs. So you'd have a pipeline component like:

```def mark_non_entities_as_unknown(doc):
    '''After an NER model has run, unset its O predictions so that a subsequent model can predict those words.'''
    ar = doc.to_array([ENT_IOB])
    ar[ar == 2] = 0 # Trusting this is right? numpy indexing confuses me.
    doc = doc.from_array([ENT_IOB], ar.reshape((ar.shape[0],1)))
    return doc

nlp.add_pipe(mark_non_entities_as_unknown, before='ner2')
```

**Comment by AIRLegend** [comment]
Added your code after the training phase (I suppose 'ner2' == 'ner') so I have the following pipes:
`['tagger', 'parser', 'mark_non_entities_as_unknown', 'ner']`

It now recognizes  the default entities again, but 'PAIN' remains as *NOT_ENTITY*. Have I done something wrong in the training part?

Thanks
